arbor/include/arbor/simd/sve.hpp

```diff
@@ -0,0 +1,900 @@
+        svuint32_t xhi = svreinterpret_u32_u64(xt);
+        auto emask = svdup_n_u32(0x7ff00000);
+        auto ebiased = svlsr_n_u32_z(svptrue_b64(), svand_u32_z(svptrue_b64(), xhi, emask), 20);
+
+        return svsub_s32_z(svptrue_b64(), svreinterpret_s32_u32(ebiased), svdup_n_s32(1023));
+    }
+
+    static svfloat64_t fraction_normal(const svfloat64_t& x) {
+        svuint64_t emask = svdup_n_u64(-0x7ff0000000000001);
+        svuint64_t bias =  svdup_n_u64(0x3ff0000000000000);
+        return svreinterpret_f64_u64(
+            svorr_u64_z(svptrue_b64(), bias, svand_u64_z(svptrue_b64(), emask, svreinterpret_u64_f64(x))));
+    }
+
+    static inline svfloat64_t horner1(svfloat64_t x, double a0) {
+        return add(x, broadcast(a0));
+    }
+
+    static inline svfloat64_t horner(svfloat64_t x, double a0) {
+        return broadcast(a0);
+    }
+
+    template <typename... T>
+    static svfloat64_t horner(svfloat64_t x, double a0, T... tail) {
+        return fma(x, horner(x, tail...), broadcast(a0));
+    }
+
+    template <typename... T>
+    static svfloat64_t horner1(svfloat64_t x, double a0, T... tail) {
+        return fma(x, horner1(x, tail...), broadcast(a0));
+    }
+
+    static svfloat64_t fms(const svfloat64_t& a, const svfloat64_t& b, const svfloat64_t& c) {
+        return svnmsb_f64_z(svptrue_b64(), a, b, c);
+    }
+};
+
+}  // namespace detail
+
+namespace simd_abi {
+template <typename T, unsigned N> struct sve;
+template <> struct sve<double, 0> {using type = detail::sve_double;};
+template <> struct sve<int, 0> {using type = detail::sve_int;};
+
+template <typename T> struct reg_type;
+template <> struct reg_type<int> { using type = svint64_t;};
+template <> struct reg_type<double> { using type = svfloat64_t;};
+
+template <typename T> struct mask_type;
+template <> struct mask_type<int> { using type = svbool_t;};
+template <> struct mask_type<double> { using type = svbool_t;};
+
+template<typename Type> struct type_to_impl;
+template<> struct type_to_impl<svint64_t> { using type = detail::sve_int;};
+template<> struct type_to_impl<svfloat64_t> { using type = detail::sve_double;};
+template<> struct type_to_impl<svbool_t> { using type = detail::sve_mask;};
+};  // namespace simd_abi
+
+using namespace arb::simd::simd_abi;
+using arb::simd::detail::simd_traits;
+
+template <typename Value, unsigned N, template <class, unsigned> class Abi>
+struct simd_wrap;
+
+template <typename Value, template <class, unsigned> class Abi>
+struct simd_wrap<Value, (unsigned)0, Abi> { using type = typename simd_abi::reg_type<Value>::type; };
+
+template <typename Value, unsigned N, template <class, unsigned> class Abi>
+struct simd_mask_wrap;
+
+template <typename Value, template <class, unsigned> class Abi>
+struct simd_mask_wrap<Value, (unsigned)0, Abi> { using type = typename simd_abi::mask_type<Value>::type; };
+
+// Math functions exposed for SVE types
+
+#define SVE_UNARY_ARITHMETIC(name)\
```

Comment:
As these macros are for internal use but populate the global namespace, I'd prefix them with `ARB_` to help avoid collisions, and put a `_` suffix on the end, to indicate that they are only used internally. They can in fact be `#undef`ed after their use is finished.

Reply:
Okay